### PR TITLE
 BUGFIX | Teleport ALPN Proxy doesn't respect HTTP CONNECT Proxy

### PR DESCRIPTION
**Pull request by alexatcanva**
# Intent

This change allows the generated ALPN proxy to honour the HTTP_PROXY environment variable when specified

# Background

If a user was interacting with Teleport Proxy that was placed behind a HTTP CONNECT proxy. The HTTPS_PROXY flag was not honoured.
This resulted in connections potentially timing out when an engineer was using tools such as tsh db connect ...

This change resolves this issue by calling the NewDialer from github.com/gravitational/teleport/api/client

Closing #17375 in favour of this PR

### CODE REVIEW

**Comment by zmb3** [comment]
Thanks @alexatcanva!

@smallinsky would you mind taking a look?

**Comment by alexatcanva** [comment]
Hey @zmb3 and @smallinsky In the other PR I was branching from Teleport V8, which I readily have development infrastructure to test this proxy implementation, Can I please ask to for a hand with testing this PR, as I don't have the available development infrastructure to accurately test this change (Teleport v12(?)). 🙏 

Let me know if there's anything else I can do to help out!

File: lib/srv/alpnproxy/dialer.go [hunk 1]
```diff
@@ -64,10 +65,7 @@ func (d ALPNDialer) DialContext(ctx context.Context, network, addr string) (net.
 		return nil, trace.BadParameter("missing TLS config")
 	}
 
-	var dialer ContextDialer = &net.Dialer{
-		KeepAlive: d.cfg.KeepAlivePeriod,
-		Timeout:   d.cfg.DialTimeout,
-	}
+	var dialer ContextDialer = apiclient.NewDialer(ctx, d.cfg.KeepAlivePeriod, d.cfg.DialTimeout)
 	if d.cfg.ALPNConnUpgradeRequired {
 		dialer = newALPNConnUpgradeDialer(d.cfg.KeepAlivePeriod, d.cfg.DialTimeout, d.cfg.TLSConfig.InsecureSkipVerify)
```
Comment (smallinsky) [inline lib/srv/alpnproxy/dialer.go:70]: The `d.cfg.ALPNConnUpgradeRequired` flow will overwrite the `apiclient.NewDialer` 
I think that we cloud pass the `dialer` to `newALPNConnUpgradeDialer` and assign to https://github.com/gravitational/teleport/blob/0ee91f6c37eb3a8c5bc98b15f857c95902df50b2/lib/srv/alpnproxy/conn_upgrade.go#L88 to support custom HTTPS proxy in `ALPNConnUpgradeRequired` flow. 

@greedy52 Could you also take a look ?

Comment (greedy52) [reply lib/srv/alpnproxy/dialer.go:70]: Yes, i believe the connection upgrade flow needs fix.

If we are going to support other protocols for connection upgrades, ideally we should move the connection upgrade logic inside the api dialer, something like:
```
apiclient.NewDialer(&apiclient.DialerConfig{
  KeepAlivePeriod: ...,
  DialTiemout: ...,
  ALPNUpgradeRequired: ...,
  Insecure: ...,
})
```
Seems too big of a change for now.

So agreed with what @smallinsky suggested to replace the `net.Dialer` with api dialer inside `alpnConnUpgradeDialer` for now.

Comment (alexatcanva) [reply lib/srv/alpnproxy/dialer.go]: Hey @greedy52 and @smallinsky, I've updated the code to pass in the `apiclient.ContextDialer`, and amended and passed the effected tests. Feel free to take a look when you get a chance 🙏 

A quick question, in [dailer.go:29](https://github.com/gravitational/teleport/blob/40a4d8946a7aec1be676233b24fbe87a6fd3a289/lib/srv/alpnproxy/dialer.go#L29) we are re-defining another ContextDialer interface, and now that `apiclient` is imported we have two interfaces of the same name (although different namespace/package name) with identical signatures, is this something that we think we can refactor?

Comment (alexatcanva) [reply lib/srv/alpnproxy/dialer.go]: I'm also AFK for the next week or so so I won't be able to commit any changes til I'm back, but please feel free to leave some more review comments that I can address once I'm back! 

Thanks Team!

**Review by greedy52 — APPROVED** [review]

**Review by smallinsky — APPROVED** [review]

File: lib/srv/alpnproxy/dialer.go (hunk 1 above)
Comment (greedy52) [reply lib/srv/alpnproxy/dialer.go]: > A quick question, in [dailer.go:29](https://github.com/gravitational/teleport/blob/40a4d8946a7aec1be676233b24fbe87a6fd3a289/lib/srv/alpnproxy/dialer.go#L29) we are re-defining another ContextDialer interface, and now that apiclient is imported we have two interfaces of the same name (although different namespace/package name) with identical signatures, is this something that we think we can refactor?

Yes, I think we can use the same one from apiclient (or maybe redefine like `type ContextDialer apiclient.ContextDialer`)

**Comment by alexatcanva** [comment]
Heya @smallinsky 👋 let me know if there's anything else I can do to help get this merged 🙏

File: lib/srv/alpnproxy/conn_upgrade.go [hunk 2]
```diff
@@ -20,17 +20,15 @@ import (
...
-	"github.com/gravitational/trace"
-	"github.com/sirupsen/logrus"
-
 	"github.com/gravitational/teleport"
+	apiclient "github.com/gravitational/teleport/api/client"
 	"github.com/gravitational/teleport/api/defaults"
 	"github.com/gravitational/teleport/lib/srv/alpnproxy/common"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"net"
```
Comment (smallinsky) [inline lib/srv/alpnproxy/conn_upgrade.go:29]: nit: could you group imports.

File: lib/srv/alpnproxy/conn_upgrade_test.go [hunk 3]
```diff
@@ -21,6 +21,7 @@ import (
 	"crypto/tls"
 	"crypto/x509/pkix"
 	"errors"
+	apiclient "github.com/gravitational/teleport/api/client"
```
Comment (smallinsky) [inline lib/srv/alpnproxy/conn_upgrade_test.go:24]: nit: could you group imports.

**Review by smallinsky — COMMENTED** [review]
@alexatcanva Thanks for the contribution. I left one nit comment regraind go import grouping but apart from that LGTM. 

 @zmb3 @r0mant
Seems that your approvals are needed to merge. Could you take a look and stamp this PR ?

**Review by zmb3 — APPROVED** [review]

File: lib/srv/alpnproxy/conn_upgrade_test.go (hunk 3 above)
Comment (zmb3) [reply lib/srv/alpnproxy/conn_upgrade_test.go:24]: `make fix-imports` will fix this up and make the linter happy

**Comment by alexatcanva** [comment]
@zmb3 ran the `fix-imports` let me know if there's anything else! :D

**Comment by zmb3** [comment]
Looks good, thanks @alexatcanva. I'll give this one more try with GCB and if it fails I will buddy your changes into a separate PR. GCB sometimes gets weird when the PR comes from a fork.

**Review by r0mant — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Comment by zmb3** [comment]
@smallinsky @r0mant I've opened a buddy at #18993, please approve there and we'll get this merged.

**Pull request closed without merging**